BuildSourceImage.sh

```diff
@@ -1,80 +1,968 @@
+    echo -n ${#c}
+}
+
+# size of file in bytes
+_size() {
+    local file="${1}"
+    stat -c "%s" "${file}" | tr -d '\n'
+}
+
+# date timestamp in RFC 3339, to the nanosecond, but slightly golang style ...
+_date_ns() {
+    date --rfc-3339=ns | tr ' ' 'T' | tr -d '\n'
+}
+
+# local `mktemp -d`
+_mktemp_d() {
+    mktemp -d "${TMPDIR:-/tmp}/${ABV_NAME}.XXXXXX"
+}
+
+# local `mktemp`
+_mktemp() {
+    mktemp "${TMPDIR:-/tmp}/${ABV_NAME}.XXXXXX"
+}
+
+# local rm -rf
+_rm_rf() {
+    _debug "rm -rf $@"
+    rm -rf $@
+}
+
+# local mkdir -p
+_mkdir_p() {
+    if [ -n "${DEBUG}" ] ; then
+        mkdir -vp $@
+    else
+        mkdir -p $@
+    fi
+}
+
+# local tar
+_tar() {
+    if [ -n "${DEBUG}" ] ; then
+        tar -v $@
+    else
+        tar $@
+    fi
+}
+
+# output things, only when $DEBUG is set
+_debug() {
+    if [ -n "${DEBUG}" ] ; then
+        echo "[${ABV_NAME}][DEBUG] ${@}"
+    fi
+}
+
+# general echo but with prefix
+_info() {
+    echo "[${ABV_NAME}][INFO] ${@}"
+}
+
+_warn() {
+    echo "[${ABV_NAME}][WARN] ${@}" >&2
+}
+
+# general echo but with prefix
+_error() {
+    echo "[${ABV_NAME}][ERROR] ${@}" >&2
+    exit 1
+}
+
+#
+# parse the OCI image reference, accounting for:
+# * transport name
+# * presence or lack of transport port number
+# * presence or lack of digest
+# * presence or lack of image tag
+#
+
+#
+# return the image reference's digest, if any
+#
+parse_img_digest() {
+    local ref="${1}"
+    local digest=""
+    if [ "$(_count '@' ${ref})" -gt 0 ] ; then
+        digest="${ref##*@}" # the digest after the "@"
+    fi
+    echo -n "${digest}"
+}
+
+#
+# determine image base name (without tag or digest)
+#
+parse_img_base() {
+    local ref="${1%@*}" # just the portion before the digest "@"
+    local base="${ref}" default to the same
+    if [ "$(_count ':' $(echo ${ref} | tr '/' '\n' | tail -1 ))" -gt 0 ] ; then
+        # which means everything before it is the base image name, **including
+        # transport (which could have a port delineation), and even a URI
+        base="$(echo ${ref} | rev | cut -d : -f 2 | rev )"
+    fi
+    echo -n "${base}"
+}
+
+#
+# determine, or guess, the image tag from the provided image reference
+#
+parse_img_tag() {
+    local ref="${1%@*}" # just the portion before the digest "@"
+    local tag="latest" # default tag
+
+    if [ -z "${ref}" ] ; then
+        echo -n "${tag}"
+        return 0
+    fi
+
+    if [ "$(_count ':' $(echo ${ref} | tr '/' '\n' | tail -1 ))" -gt 0 ] ; then
+        # if there are colons in the last segment after '/', then get that tag name
+        tag="$(echo ${ref} | tr '/' '\n' | tail -1 | cut -d : -f 2 )"
+    fi
+    echo -n "${tag}"
+}
+
+#
+# an inline prefixer for containers/image tools
+#
+ref_prefix() {
+    local ref="${1}"
+
+    # get the supported prefixes of the current version of skopeo
+    IFS=", "
+    local pfxs=( $(skopeo copy --help | grep -A1 "Supported transports:" | grep -v "Supported transports") )
```

Comment:
```
In BuildSourceImage.sh line 183:
    local pfxs=( $(skopeo copy --help | grep -A1 "Supported transports:" | grep -v "Supported transports") )
                 ^-- SC2207: Prefer mapfile or read -a to split command output (or quote to avoid splitting).
```